[Diagnostics] Port the "extra address of argument" diagnostic to the new framework.

hborla opened the pull request.

hborla committed:
[Diagnostics] Port the "extra address of agrument" diagnostic to the new
framework.

hborla commented:
@swift-ci please smoke test

xedin reviewed (changes requested):
We already have a ConstraintFix/FailureDiagnostic for this - `RemoveAddressOf`/`InvalidUseOfAddressOf` looks like we just need to extend its use.

hborla committed:
[Diagnostics] Unify `DiagnosticFailure` and `ConstraintFix` classes for extra `&` failures.

Make `InvalidUseOfAddressOf` a `ContextualFailure`, make `ReturnAddressOf`
a `ContextualMismatch`, and extend this failure to cover using `&` with a
non-inout argument.

hborla commented:
@swift-ci please smoke test

xedin commented on lib/Sema/CSFix.h:
```diff
@@ -58,8 +58,6 @@ enum class FixKind : uint8_t {
   /// Introduce a '&' to take the address of an lvalue.
   AddressOf,
-  /// Remove extraneous use of `&`.
-  RemoveAddressOf,
```
Although this is a contextual failure, let's keep it's special kind since we know what it is.

xedin commented on lib/Sema/CSFix.cpp:
```diff
@@ -114,6 +114,17 @@ AddAddressOf *AddAddressOf::create(ConstraintSystem &cs, Type argTy,
   return new (cs.getAllocator()) AddAddressOf(cs, argTy, paramTy, locator);
 }
 
+bool RemoveAddressOf::diagnose(Expr *root, bool asNote) const {
```
Is it necessary to move stuff in `cpp` it same way as in the header?

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -908,6 +908,30 @@ bool MissingAddressOfFailure::diagnoseAsError() {
   return true;
 }
 
+SourceLoc InvalidUseOfAddressOf::getLoc() const {
```
Same thing as in `CSFix.cpp` I think it should be left in place, it might make it more difficult to blame the changes later on.

hborla commented:
@swift-ci please smoke test

xedin reviewed (approved):
Looks great, thank you!
